ide/coqide/coqOps.ml

```diff
@@ -14,7 +14,35 @@ open Ideutils
 open Interface
 open Feedback
 
-let b2c = byte_off_to_buffer_off
+let last_b2c_offsets = ref (0,0)
```

Comment:
I've been pondering about this fix, and I think it goes the wrong direction. Using a reference here is completely unsound, because we have no guarantee that the document will not change between calls to the conversion function. Instead, we should probably cache the unicode and byte locations inside the sentence type and invalidate the cache when the sentence is modified.

Reply:
Isn't the sentence is discarded when it becomes unprocessed?  If so, that should be sufficient invalidation.  And otherwise `SentenceId.changed` could reset the saved offsets to (0,0).

Reply:
You only get messages on items that are being processed, which are not modifiable.  If you later modify the buffer before that point and step forward, the new offset will be less than the last cached offset ignoring the 50 byte fudge factor: `s_byte + 50`.

Another approach, simpler to reason about, is to track the last unmodifiable location in the file.  When that's set to something less than the cached buffer offset, clear `last_b2c_offsets`.  I added code to do this; performance is the same.

WDYT?

Reply:
I don't know why you really insist to have a global cache + some hacks to make it resilient when this is clearly very fragile (and I am still unconvinced this is sound, you can probably break it with the invalid proof editing). When building an element of the sentence type you already have to compute the equivalence byte-UTF8, just store it there so you get the data locally, and attach a signal to the starting mark so that if it moves the cache is invalidated.

Reply:
Specifically, we could move the equivalence computation to `fill_command_queue`, which would be better than where it is now because we have the invariant that the buffer processing is atomic (no Coq calls can occur). Just tweak `mk_sentence` to store the byte offset in addition to the iter offset, and hook onto the mark signal for invalidation.

Reply:
The signal is not essential.  `mk_sentence` can save the initial offset for the mark in addition to `(s_uni, s_byte)`.  Then `loc_to_uni` can check whether those two differ and, if so, recompute `(s_uni, s_byte)`.  To do that efficiently, the code would step backward through the sentences of the document until it finds a sentence that still has valid offsets.  At that point it can proceed forward, updating the offsets for all the examined sentences with invalid offsets.  Offsets for sentence N would be computed using the offsets for sentence N-1 as the starting point.  This would be O(S + C) where S is the number of sentences and C is the number of characters in the buffer.  Updates could be done lazily (i.e. only when `loc_to_uni` is called).

If you like this, I'll code it.  I don't think it qualifies as simpler.

Reply:
This is still unsound, because an unchanged offset doesn't mean that the byte location didn't change. For instance, if you replace one UTF8 character that fits in one byte (i.e. an ASCII code) by one that fits in several bytes, then you'll get a corrupted cache without a way to realize it. I agree that it's probably fairly uncommon, so maybe it's worth implementing for 8.15.1 as a hotfix and refine it afterwards.

Reply:
(I have started implementing such a fix a few days ago, maybe I'll just push it as a new PR if I can polish it.)

Reply:
That's #15820.